### PR TITLE
HHH-9342 HQL "x member of treat(y as Type).collection" fails to parse

### PR DESCRIPTION
**Pull request by EtienneMiret**
See [HHH-9342](https://hibernate.atlassian.net/browse/HHH-9342).

I added a new construction "treatAsPath" in the grammar and used it in the "member of" rule. This solves the specific bug I found. However I’m pretty sure this new construction should be used at other places too.

### CODE REVIEW

**Comment by brmeyer** [comment]
@sebersole, up for reviewing this one?  I need to stop putting off learning Antlr...

**Comment by sebersole** [comment]
There is already a test for TREAT handling, 2 in fact (1 in hibernate-core and 1 in hibernate-entitymanager):

hibernate-core - org.hibernate.test.jpa.ql.TreatKeywordTest
hibernate-entitymanager - org.hibernate.jpa.test.criteria.TreatKeywordTest

**Comment by sebersole** [comment]
I have tried expanding the visibility of TREAT in HQL grammars, but the problem is that it always leads to ambiguities.  This one works because it is very limited in scope/visibility.

I'll look at this some more.

File: hibernate-core/src/main/antlr/hql.g [hunk 1]
```diff
@@ -846,6 +846,13 @@ path
 	: identifier ( DOT^ { weakKeywords(); } identifier )*
 	;
 
+treatAsPath
+	: path
+	| i:IDENT! OPEN! p:treatAsPath AS! a:path! CLOSE! ( DOT^ path )? {i.getText().equalsIgnoreCase("treat") }? {
```
Comment (sebersole) [inline hibernate-core/src/main/antlr/hql.g]: TREATs cannot be nested.  

Also, not a fan of recognizing IDENT and then applying a semantic predicate to recognize a specific IDENT.  Why did you do that as opposed to just recognizing "treat"?

Comment (sebersole) [inline hibernate-core/src/main/antlr/hql.g]: Oh, nevermind about the IDENT + semantic predicate.  We did that to avoid making TREAT a keyword.

**Comment by sebersole** [comment]
This will not apply cleanly for me to master, so I am going to have to manually apply the grammar change.  I'll also manually apply the test method to the proper existing class

**Comment by sebersole** [comment]
Rather than a validating semantic predicate, I flipped it and used a gated semantic predicate (with look-ahead) instead to distinguish between the 2 paths.

**Comment by sebersole** [comment]
Applied equivalent changes manually upstream as this would not apply cleanly.

**Pull request closed without merging**